Python/Module5_OddsAndEnds/WorkingWithFiles.md

```diff
@@ -249,6 +249,102 @@ with open("a_poem.txt", mode="r") as my_open_file:
+
+>>> my_data.shape
+(numSamples, numAttributesPerSample)
+
+#Let's look at the actual data
+>>> my_data
+array([[    nan,     nan,     nan, ...,     nan,     nan,     nan],
+       [    nan, -99.9  , -99.9  , ..., -99.9  , -99.9  , -99.9  ],
+       [    nan,   0.875,   1.39 , ...,   4.506, -99.9  , -99.9  ],
+       ...,
+       [    nan,   2.157,   3.43 , ...,  12.89 ,  97.   ,  21.95 ],
+       [    nan,   2.087,   2.84 , ...,  10.963,  92.   ,  21.95 ],
+       [    nan,   1.926,   2.98 , ...,  12.228,  84.   ,  21.95 ]])
+```
+You may notice that there are some `nan` values present when we look at this perticular set of data. Typically, if there are non-numerical values in the file, such as headers and dates, importing it into a Numpy array will turn those values into `nan`.
+
+### How to parse CSVs with Pandas
+
+A really popular library for parsing through CSVs is the [Pandas](https://pandas.pydata.org/pandas-docs/stable/index.html "Pandas Documentation") library. Here is a quick way to parse through a CSV using Pandas:
+```Python
+import pandas as pd
+my_data =pd.read_csv(r"path_to_csv/file.csv", sep=',',header=None)
+```
+That's it! The method `read_csv()` imports the CSV into the variable `my_data`. This method has similar input parameters to `genfromtxt()` and many extra optional parameters as well. Look at the docstring to see more.
+
+ Let's look at some properties of an [example file](https://www.kaggle.com/account/login?returnUrl=%2Fjolasa%2Fwaves-measuring-buoys-data-mooloolaba%2Fversion%2F1 "Kaggle Dataset") that has data related to ocean waves:
+
+```Python
+>>> type(my_data)
+pandas.core.frame.DataFrame #Notice that this is a custom type
+
+>>> my_data.shape
+(numSamples, numAttributesPerSample)
+
+>>> my_data.values #This is how we access the values as an array
+array([['Date/Time', 'Hs', 'Hmax', ..., 'Tp', 'Peak Direction', 'SST'],
+       ['01/01/2017 00:00', '-99.9', '-99.9', ..., '-99.9', '-99.9',
+        '-99.9'],
+       ['01/01/2017 00:30', '0.875', '1.39', ..., '4.506', '-99.9',
+        '-99.9'],
+       ...,
+       ['30/06/2019 22:30', '2.157', '3.43', ..., '12.89', '97', '21.95'],
+       ['30/06/2019 23:00', '2.087', '2.84', ..., '10.963', '92',
+        '21.95'],
+       ['30/06/2019 23:30', '1.926', '2.98', ..., '12.228', '84',
+        '21.95']], dtype=object)
+```
+One of the coolest features of Pandas is how it nicely organizes the parsed CSV data for visualization. Here is how `my_data` is displayed in a jupyter notebook:
+```Python
+>>> my_data[0:21] #Prints out first 20 values in nice format
+```
+![Memory consumption figure](pics/Pandas_CSV.jpg)
+
+One of the main advantages of Pandas is that it **treats all the data as strings**, while Numpy only deals with numerical values. This allows Pandas to store information such as headers and date, while Numpy cannot. Read the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/index.html "Documentation Link") for more information.
```

Comment:
Is treating everything as a string an advantage? Maybe something to be aware of?

Stylize NumPy